[Local GC] Add a Standalone GC loader design document

swgillespie opened the pull request:
[Rendered View](https://github.com/swgillespie/coreclr/blob/c59801e09c0bdb13dccf3cafbb320e3c74722c63/Documentation/design-docs/standalone-gc-loading.md)

This PR introduces a design document for the process of loading and initializing a standalone GC. It is styled after the way that the JIT loads standalone JITs and it does not deviate much from the prior art.

Design reviews and comments greatly appreciated!

mikedn commented on Documentation/design-docs/standalone-gc-loading.md:
```diff
@@ -0,0 +1,222 @@
+
+It is not an error to emit either `StandaloneGCLocation` or `StandaloneGCName`. If not specified, and
+`UseStandaloneGC` is set, `StandaloneGCLocation` will default to the same directory as CoreCLR and `StandaloneGCName`
+will default to `gc` (`gc.dll` on windows, `libgc.so/dylib` on Linux/OSX).
```
Perhaps the default library name should be `clrgc`, not just `gc`?

swgillespie replied:
Good point, especially to avoid conflicts with the BDW GC.

swgillespie replied:
I think I'm going to remove this part from the design doc, but we definitely should be using `clrgc` as the default name in the implementation.

jkotas commented on Documentation/design-docs/standalone-gc-loading.md:
```diff
@@ -0,0 +1,222 @@
+
+Author: Sean Gillespie (@swgillespie) - 2017
+
+A standalone GC (or "local GC", as it is often called) initiaive has
```
This may rather be in PR description. Not in the design document that should stay relevant over time.

swgillespie replied:
ok!

jkotas commented on Documentation/design-docs/standalone-gc-loading.md:
```diff
@@ -0,0 +1,222 @@
+  will be trivially portable to other runtime implementations.
+
+  This is a significant engineering pain point today. A great deal of time is spent manually porting
+  changes between the different branches of the GC codebase. When changes are not ported, the GC
```
FWIW, you still need to do the work to flow the changes between different products/branches/releases even with standalone GC. It may be easier with enforced interface, but it is still going to take a great deal of time. 

In the RyuJIT case, the maintenance of JIT/EE interface and SuperPMI is non-trivial effort.

swgillespie replied:
fair enough, I did not mean to imply that. I'll change the wording.

jkotas commented on Documentation/design-docs/standalone-gc-loading.md:
```diff
@@ -0,0 +1,222 @@
+
+There are three configuration keys that control the behavior of loading a standalone GC:
+
+1. `UseStandaloneGC`, (environment variable `COMPLUS_UseStandaloneGC`). If set and equal to "1", the EE
```
Why do we need 3 things? Can we have just one? `COMPlus_GCName`?

I believe that it is how the JIT loading works - there is a single key `COMPlus_JitName` that you set, and it gets passed to LoadLibrary. Nothing fancy.

swgillespie replied:
Sure. I modeled it with this comment in mind: https://github.com/dotnet/coreclr/blob/2f37c02dfa3b649d4d7322724c95ab08cb12cb6f/src/vm/codeman.cpp#L1536-L1537. Keeping it simple sounds good to me.

jkotas replied:
`g_CLRJITPath` is meant for loading the JIT from a nuget package: You reference a nuget package with a different JIT in your project and it gets pulled in. It required changes in number of places across the board to make it happen... .

swgillespie replied:
I see - thanks!

jkotas commented on Documentation/design-docs/standalone-gc-loading.md:
```diff
@@ -0,0 +1,222 @@
+
+There are three configuration keys that control the behavior of loading a standalone GC:
+
+1. `UseStandaloneGC`, (environment variable `COMPLUS_UseStandaloneGC`). If set and equal to "1", the EE
```
Nit: `COMPLUS` should be spelled as `COMPlus`. The casing matters on Unix.

swgillespie replied:
ok!

jkotas commented on Documentation/design-docs/standalone-gc-loading.md:
```diff
@@ -0,0 +1,222 @@
+
+### Loader State Machine
+
+The loading of a standalone GC by the EE is governed by a state machine with these states:
```
The state machine makes sense if the set of transitions between different states is non-trivial. If the only way is from one state to the next with no forks or back edges, it is overkill to think about this as a state machine.

The sequence is described above. I do not think it needs to be repeated here. It is implementation detail.

swgillespie replied:
Makes sense. Perhaps it should be phrased more as a "status", like this: https://github.com/dotnet/coreclr/blob/2f37c02dfa3b649d4d7322724c95ab08cb12cb6f/src/vm/codeman.cpp#L1455. I was hoping for something where the point of failure is easily observed in a debugger or dump.

jkotas commented on Documentation/design-docs/standalone-gc-loading.md:
```diff
@@ -0,0 +1,222 @@
+
+The question of whether or not the EE should attempt to locate and load a standalone GC
+is answered by the EE's configuration system (`EEConfig`). EEConfig has the ability to
+query configuration information from a variety of sources, such a registry keys or
```
Registry keys are read in .NET Framework only, not in .NET Core.

swgillespie replied:
ok!

noahfalk commented on Documentation/design-docs/standalone-gc-loading.md:
```diff
@@ -0,0 +1,222 @@
+How can we provide the most useful error message when a standalone GC fails to load? In the past it has been difficult
+to determine what preciscely has gone wrong with `coreclr_initialize` returns a HRESULT and no indication of what occured.
+
+Same question for the DAC - Is `E_FAIL` the best we can do? If we could define our own error for DAC/GC version
```
I would encourage a more specific HRESULT. Lets use a pragmatic definition of breaking change here - do SOS/ICorDebug tests behave worse with E_SPECIFIC_ERROR than E_FAIL? If so we should re-evaluate but I suspect it will make little difference. My hunch is the bigger issue will be getting tools to fail cleanly when GC makes a breaking change regardless of the HRESULT you use to report the error.

swgillespie replied:
sounds good! I'll take a look at this in detail today.

swgillespie replied:
SOS would not behave any differently if a failure HRESULT other than `E_FAIL` is returned. I did find this: https://github.com/dotnet/coreclr/blob/master/src/debug/di/process.cpp#L205-L207 - I suppose we would need to add a new error code tot his check or emit `CORDB_E_UNCOMPATIBLE_PLATFORMS` to avoid trouble.

noahfalk replied:
Adding a new HRESULT to the di/process.cpp would be my preference. CORDB_UNCOMPATIBLE_PLATFORMS specifically means a difference in hardware architecture and I'd rather not stretch its meaning to include versioning of local GC.

Did you get a chance to see observe the user-visible behavior you get in common tools when do return a failing HRESULT (even if it is E_FAIL)?

swgillespie replied:
I've looked at CLRMD and mdbg. CLRMD does not attempt to handle any `COMException` so the buck is passed on to users. Mdbg does handle `COMException`s and, in particular, special-cases E_FAIL in a couple places that I've found to do something other than propagate the error:

* Retrieving the current exception on a managed thread
* Retrieving the current notification on a thread
* Trying to resolve a source location into an IL location
* Trying to find the entry point of an assembly
* Trying to find symbols for a method

None of these would be affected by a non-E_FAIL HRESULT from GC-related DBI methods.

I haven't looked at clrdbg since I'm not sure where the source for that is.

noahfalk replied:
clrdbg is a pretty large codebase so I'm not sure how successful you would be at trying to do code review. I'd suggest directly running Visual Studio debugger and seeing what behavior you see when you try to debug an application with a local GC that reports an incompatible version. Assuming that basic breakpoints, stepping, callstacks, locals all work I would probe more on the UI for taking heap snapshots and diffing them. I would also check out what it looks like to use windbg+SOS, in particular the SOS commands that examine the GC heap.

swgillespie replied:
Ok! I haven't yet implemented this feature, so would you be OK with keeping this document as-is and then amending it once I have the opportunity to test clrdbg in this manner?

noahfalk replied:
I have no qualms about the doc content, feel free to merge at any time.

swgillespie replied:
great, thanks!

swgillespie commented:
Just pushed a new version to address some feedback - now I'm looking at SOS and the DBI to see how they treat `E_FAIL`.

swgillespie commented:
I believe I've addressed all of the feedback I've gotten so far - if there aren't any further objections, I'm ready to merge this and get started on the implementation.

sergiy-k commented:
LGTM

swgillespie commented:
thanks! I'll get started 😄
